<a href="https://colab.research.google.com/github/ukhyun-lee/stock_market_analysis/blob/main/2%EC%A3%BC%EC%B0%A8/pandas_exercise.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

np.random.seed(0)

# 과제 1
## Q1. 


<div class="alert alert-success">
[문제1]     
    
insurance.csv'파일 읽어서 df에 할당하고, 상위 5개 항목 출력하기
</div>

------------

[힌트]

pd.read_csv()

In [3]:
 df = pd.read_csv('https://raw.githubusercontent.com/zzhining/ml_basic/main/dataset/insurance.csv')
 df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

## Q2. 


<div class="alert alert-success">
[문제2]     
    
df의 요약(평균, 4분위수, 전체 갯수 등)정보 확인하기
</div>

------------

In [4]:
df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

## Q3. 


<div class="alert alert-success">
[문제3]     
    
신규 컬럼(double bmi)을 생성하고, bmi의 2배가 되는 값을 할당하기
</div>

------------

In [11]:
df['doublebmi'] = df['bmi'] * 2
df.head()

age     sex     bmi  children smoker     region      charges  deft  \
0   19  female  27.900         0    yes  southwest  16884.92400     0   
1   18    male  33.770         1     no  southeast   1725.55230     1   
2   28    male  33.000         3     no  southeast   4449.46200     2   
3   33    male  22.705         0     no  northwest  21984.47061     3   
4   32    male  28.880         0     no  northwest   3866.85520     4   

   doublebmi  
0      55.80  
1      67.54  
2      66.00  
3      45.41  
4      57.76

## Q4. 


<div class="alert alert-success">
[문제4]     
    
신규 컬럼(debt)을 생성하고, 0으로 채우기
</div>

------------

In [13]:
df['deft'] = 0
df.head()

age     sex     bmi  children smoker     region      charges  deft  \
0   19  female  27.900         0    yes  southwest  16884.92400     0   
1   18    male  33.770         1     no  southeast   1725.55230     0   
2   28    male  33.000         3     no  southeast   4449.46200     0   
3   33    male  22.705         0     no  northwest  21984.47061     0   
4   32    male  28.880         0     no  northwest   3866.85520     0   

   doublebmi  
0      55.80  
1      67.54  
2      66.00  
3      45.41  
4      57.76

## Q5. 


<div class="alert alert-success">
[문제5]     
    
debt에 0부터 값이 증가하도록 채우기
</div>

------------

In [8]:
range(10)
df['deft'] = np.arange(len(df))
df.head()

age     sex     bmi  children smoker     region      charges  deft
0   19  female  27.900         0    yes  southwest  16884.92400     0
1   18    male  33.770         1     no  southeast   1725.55230     1
2   28    male  33.000         3     no  southeast   4449.46200     2
3   33    male  22.705         0     no  northwest  21984.47061     3
4   32    male  28.880         0     no  northwest   3866.85520     4

## Q6. 


<div class="alert alert-success">
[문제6]     
    
1번 인덱스의 정보 출력
</div>

------------

[힌트]
iloc



age                  18  
sex                male  
bmi               33.77  
children              1  
smoker               no  
region        southeast  
charges         1725.55  
double bmi        67.54  
debt                  1  

In [15]:
df.head()
df.iloc[1]

age                 18
sex               male
bmi              33.77
children             1
smoker              no
region       southeast
charges      1725.5523
deft                 0
doublebmi        67.54
Name: 1, dtype: object

## Q7. 


<div class="alert alert-success">
[문제7]     
    
'double bmi'열 삭제하기
</div>

In [22]:
# axis = 1:  축을 컬럼방향으로
# inplace = True :  df객체에 있는 값을 연산 결과로 대체
df.drop(['doublebmi'], axis=1, inplace=True )

In [23]:
df.head()

age     sex     bmi  children smoker     region      charges  deft
0   19  female  27.900         0    yes  southwest  16884.92400     0
1   18    male  33.770         1     no  southeast   1725.55230     0
2   28    male  33.000         3     no  southeast   4449.46200     0
3   33    male  22.705         0     no  northwest  21984.47061     0
4   32    male  28.880         0     no  northwest   3866.85520     0

## Q8. 


<div class="alert alert-success">
[문제8]     
    
 'age'열, 'sex'열만 출력하기
</div>

In [24]:
df[['age', 'sex']]

age     sex
0      19  female
1      18    male
2      28    male
3      33    male
4      32    male
...   ...     ...
1333   50    male
1334   18  female
1335   18  female
1336   21  female
1337   61  female

[1338 rows x 2 columns]

# 과제2
## Q9.
<div class="alert alert-success">
[문제9]     
    
나이를 1살씩 증가시키기
</div>

In [32]:
df['age'] = df['age'] + 1
df.head()

age     sex     bmi  children smoker     region      charges  deft
0   20  female  27.900         0    yes  southwest  16884.92400     0
1   19    male  33.770         1     no  southeast   1725.55230     0
2   29    male  33.000         3     no  southeast   4449.46200     0
3   34    male  22.705         0     no  northwest  21984.47061     0
4   33    male  28.880         0     no  northwest   3866.85520     0

## Q10.
<div class="alert alert-success">
[문제10]     
    
charges의 표시형식을 소수점 아래 2자리로 바꾸기
</div>

[힌트]

함수 생성, 적용(apply/map)

In [34]:
# round() : 반올림
df['charges'].round(2)

0       16884.92
1        1725.55
2        4449.46
3       21984.47
4        3866.86
          ...   
1333    10600.55
1334     2205.98
1335     1629.83
1336     2007.94
1337    29141.36
Name: charges, Length: 1338, dtype: float64

In [38]:
test_num = 123.45678

def sosu(data):
    return format(data, ".2f")

sosu(345322.234534)  

'345322.23'

In [40]:
df['charges'] = df['charges'].apply(sosu)

In [41]:
df.head()

age     sex     bmi  children smoker     region   charges  deft
0   20  female  27.900         0    yes  southwest  16884.92     0
1   19    male  33.770         1     no  southeast   1725.55     0
2   29    male  33.000         3     no  southeast   4449.46     0
3   34    male  22.705         0     no  northwest  21984.47     0
4   33    male  28.880         0     no  northwest   3866.86     0

## Q11.
<div class="alert alert-success">
[문제11]     
    
컬럼 정렬(알파벳 순서대로) : age, bmi, charges... 순으로 정렬
</div>

[Hint]
sort_index()


In [43]:
df.sort_index(axis=1)

age     bmi   charges  children  deft     region     sex smoker
0      20  27.900  16884.92         0     0  southwest  female    yes
1      19  33.770   1725.55         1     0  southeast    male     no
2      29  33.000   4449.46         3     0  southeast    male     no
3      34  22.705  21984.47         0     0  northwest    male     no
4      33  28.880   3866.86         0     0  northwest    male     no
...   ...     ...       ...       ...   ...        ...     ...    ...
1333   51  30.970  10600.55         3     0  northwest    male     no
1334   19  31.920   2205.98         0     0  northeast  female     no
1335   19  36.850   1629.83         0     0  southeast  female     no
1336   22  25.800   2007.94         0     0  southwest  female     no
1337   62  29.070  29141.36         0     0  northwest  female    yes

[1338 rows x 8 columns]

## Q12.
<div class="alert alert-success">
[문제12]     
    
값 정렬(age) : 나이 순으로 정렬
</div>

[Hint]
sort_values


In [45]:
df.sort_values(by='age')

age     sex     bmi  children smoker     region   charges  deft
1248   19  female  39.820         0     no  southeast   1633.96     0
482    19  female  31.350         0     no  southeast   1622.19     0
492    19  female  25.080         0     no  northeast   2196.47     0
525    19  female  33.880         0     no  southeast  11482.63     0
529    19    male  25.460         0     no  northeast   1708.00     0
...   ...     ...     ...       ...    ...        ...       ...   ...
398    65    male  25.600         2     no  southwest  14988.43     0
335    65    male  34.500         0     no  southwest  13822.80     0
378    65  female  30.115         3     no  northwest  16455.71     0
1265   65    male  23.760         0    yes  southeast  26926.51     0
635    65    male  38.190         0     no  northeast  14410.93     0

[1338 rows x 8 columns]

## Q13.
<div class="alert alert-success">
[문제13]     
    
나이('age') 평균구하기
</div>

[Hint]
mean

In [46]:
df['age'].mean()

40.20702541106129

## Q14.
<div class="alert alert-success">
[문제14]     
    
보험료('charges')와 가장 상관관계가 높은 변수는?
</div>

[Hint]
 - 공분산: df.cov()
 - 상관계수: df.corr()

# 과제 3
## Q15.
<div class="alert alert-success">
[문제15]     
    
 남자, 여자가 각각 몇명있는지 구하기
</div>

[Hint]
pd.value_counts


## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가 3명인 데이터 출력하기
</div>

## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가  5명 이상인 데이터 출력하기
</div>

## Q17.
<div class="alert alert-success">
[문제17]     
    
 bmi가  17 미만인 데이터 출력하기
</div>

# 과제 4 
## Q18.
<div class="alert alert-success">
[문제18]     
    
성별에 따른 보험료 평균
</div>

## Q19.
<div class="alert alert-success">
[문제19]     
    
흡연여부에 따른 보험료 평균
</div>

## Q20.
<div class="alert alert-success">
[문제20]     
    
지역별 보험료 평균
</div>

## Q21.
<div class="alert alert-success">
[문제21]     
    
흡연여부 + 지역별 데이터 평균
</div>

## Q22.
<div class="alert alert-success">
[문제22]     
    
 흡연여부 + 지역별 데이터 평균에서 'charges' 만 츨력하기
</div>

## Q23.
<div class="alert alert-success">
[문제23]     
    
 흡연여부 + 지역별 데이터 크기(size)
</div>

## Q24.
<div class="alert alert-success">
[문제24]     
    
smoker 여부에 따라, group_val 컬럼에 '흡연자', '비흡연자' 표시하기
</div>